### PR TITLE
Discord RPC Plugin for OuterTune

### PR DESCRIPTION
**Pull request by ruskcoder**
I added a Discord RPC addon for OuterTune. Just like when you use Spotify connected to Discord, the app will now show what you are listening to on Discord. 

image

Just log in using Discord through the settings page, choose your options, and get listening!

image

### CODE REVIEW

**Review by ruskcoder — COMMENTED** [review]
I changed a few build.gradle file settings so the app would work on my phone, but those can be changed back. But the dependencies are needed.

File: app/build.gradle.kts [hunk 1]
```diff
@@ -114,6 +114,10 @@ ksp {
 }
 
 dependencies {
+    implementation("io.ktor:ktor-client-core:1.6.0")
+    implementation("io.ktor:ktor-client-cio:1.6.0")
```
Comment (mikooomich) [inline app/build.gradle.kts]: dependencies need to follow the existing format. define dependencies as they are in libs.version.toml, and implementation(...) here needs to reference those


Also, we already have ktor-client-core, why add it agian?

File: app/build.gradle.kts [hunk 2]
```diff
@@ -17,7 +17,7 @@ android {
 
     defaultConfig {
         applicationId = "com.dd3boh.outertune"
-        minSdk = 24
+        minSdk = 27
```
Comment (mikooomich) [inline app/build.gradle.kts:20]: Is the min sdk bump and disabling build variants needed?

File: app/build.gradle.kts [hunk 3]
```diff
@@ -46,8 +46,8 @@ android {
             reset()
 
             // all common abis
-            // include("x86_64", "x86", "armeabi-v7a", "arm64-v8a") // universal
-            isUniversalApk = false
+             include("x86_64", "x86", "armeabi-v7a", "arm64-v8a") // universal
```
Comment (mikooomich) [inline app/build.gradle.kts]: Why is this needed?

File: app/src/main/java/com/dd3boh/outertune/MainActivity.kt [hunk 4]
```diff
@@ -867,6 +875,7 @@ class MainActivity : ComponentActivity() {
                                 composable("login") {
                                     LoginScreen(navController)
                                 }
+
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/MainActivity.kt]: please remove extra new lines like this, you have a bunch of these

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                artistArtwork = json.getString("uploaderAvatar")
+
+                CoroutineScope(Dispatchers.IO).launch {
+                    try {
+                        val response: HttpResponse = clientDiscordCDN.get("https://kizzyapi-1-z9614716.deta.app/image?url=" + artistArtwork) {
+                        }
+                        artistArtworkCDN = JSONObject(response.bodyAsText()).getString("id")
+
+                    } catch (e: Exception) {
+                        println("Error: ${e.message}")
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt]: Error reports should use reportException()

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    val discordToken = ctx.dataStore.get(DiscordTokenKey, "")
+    rpc.token = discordToken
+
+    while (rpc.isRpcRunning()){
+        rpc.closeRPC()
+    }
+    val enableRPC = ctx.dataStore.get(EnableDiscordRPCKey, true)
+    val showArtistAvatar = ctx.dataStore.get(ShowArtistRPCKey, true)
+
+    if (discordToken != "" || !enableRPC){
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt]: space between closing bracket and opening curly

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+                    } catch (e: Exception) {
+                        println("Error: ${e.message}")
+                        
+                    } finally {
+                        clientDiscordCDN.close()
+                    }
+                }
+                CoroutineScope(Dispatchers.IO).launch {
+                    try {
+                        val response: HttpResponse = clientDiscordCDN.get("https://kizzyapi-1-z9614716.deta.app/image?url=" + artwork) {
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt]: blank curly bracket?

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                        }
+
+                        artworkCDN = JSONObject(response.bodyAsText()).getString("id")
+
+                        Thread.sleep(500)
+                        if (artistArtworkCDN == ""){
+                            artistArtworkCDN = "mp:external/_jGArMHI-5rpJu4qVDuiBARu8iEXnHeT0SZS6tZnZug/https/i.imgur.com/zDxXZKk.png"
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt]: Why externally fetch image? Use an icon/image/drawable as a placeholder

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+                        }
+                        rpc.setActivity(
+                                activity = Activity(
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt:103]: only single indent plz

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            } finally {
+                clientDiscordCDN.close()
+            }
+        }
+        }
+
+
+//    
+            }
+}
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt:160]: please fix indentation

File: app/src/main/java/com/dd3boh/outertune/playback/MusicService.kt [hunk 11]
```diff
@@ -434,7 +435,9 @@ class MusicService : MediaLibraryService(),
 
     private suspend fun recoverSong(mediaId: String, playerResponse: PlayerResponse? = null) {
         var playbackUrl = database.format(mediaId).first()?.playbackUrl
-
+        if (playbackUrl != null) {
+            Log.i("playbackurl", playbackUrl)
+        }
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/MusicService.kt]: no loging. If you want to have debug logging, hid it behind a debugging flag. for example: https://github.com/DD3Boh/OuterTune/blob/e82a47ead8393a21eeb95626f610432affcbb0fc/app/src/main/java/com/dd3boh/outertune/ui/utils/LocalMediaUtils.kt#L20

File: app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt [hunk 12]
```diff
@@ -82,10 +93,15 @@ class PlayerConnection(
         currentWindowIndex.value = player.getCurrentQueueIndex()
         currentMediaItemIndex.value = player.currentMediaItemIndex
         repeatMode.value = player.repeatMode
+
+
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt]: please adhere to the existing code style for this entire file.

File: app/src/main/java/com/dd3boh/outertune/ui/screens/settings/DiscordLoginScreen.kt [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+                                    println("Error: ${e.message}")
+                                } finally {
+                                    client.close()
+                                }
+                            }
+
+
+                            navController::navigateUp
+                            return true
+//                            val client = OkHttpClient.Builder()
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/ui/screens/settings/DiscordLoginScreen.kt]: why keep this?

File: app/src/main/java/com/dd3boh/outertune/ui/screens/settings/DiscordSettings.kt [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    val (showArtist, onShowArtistChange) = rememberPreference(key = ShowArtistRPCKey, defaultValue = true)
+
+
+    var isLoggedIn = remember(discordToken) {
+        discordToken != ""
+    }
+
+    Column(
+            Modifier
+                    .windowInsetsPadding(LocalPlayerAwareWindowInsets.current)
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/ui/screens/settings/DiscordSettings.kt]: indentation.

File: build.gradle.kts [hunk 15]
```diff
@@ -1,6 +1,8 @@
 plugins {
     alias(libs.plugins.hilt) apply(false)
     alias(libs.plugins.kotlin.ksp) apply(false)
+    id("io.ktor.plugin") version "2.3.12"
+
```
Comment (mikooomich) [inline build.gradle.kts]: same dependency issue here. use libs.versions.toml

File: gradle/libs.versions.toml [hunk 16]
```diff
@@ -1,5 +1,5 @@
 [versions]
-androidGradlePlugin = "8.3.2"
```
Comment (mikooomich) [inline gradle/libs.versions.toml:2]: why?

**Review by mikooomich — CHANGES_REQUESTED** [review]
Hi. Thanks for your PR. There are some issues that would need to be resolved. In general, code changes should be minimal and adhere to the existing style. Most notably spacing, indentation, dependency declaration, if/else curly spacing, etc. Auto format can help.

**Comment by ruskcoder** [comment]
Sorry about all the issues 😅 I'm pretty new to android development and I was just trying to add a feature. I'll get to work on the fixes right away!

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt (hunk 8 above)
Comment (ruskcoder) [reply app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt]: To my knowledge, Discord RPC needs to use an image from their CDN. A normal drawable/image file will need an extra HTTP request to be uploaded to the CDN. Over here, I am using a pre-uploaded image of a placeholder profile picture so I wouldn't need to make an extra request to the API.

File: gradle/libs.versions.toml (hunk 16 above)
Comment (ruskcoder) [reply gradle/libs.versions.toml:2]: I was getting this error:
The project is using an incompatible version (AGP 8.3.2) of the Android Gradle plugin. Latest supported version is AGP 8.2.2
See Android Studio & AGP compatibility options.

File: app/build.gradle.kts (hunk 2 above)
Comment (ruskcoder) [reply app/build.gradle.kts:20]: I got this error because of the KizzyRPC dependency. What should I go with? 

uses-sdk:minSdkVersion 24 cannot be smaller than version 27 declared in library [com.github.dead8309:KizzyRPC:1.0.71]

	Suggestion: use a compatible library with a minSdk of at most 24,
		or increase this project's minSdk version to at least 27,
		or use tools:overrideLibrary="com.my.kizzyrpc" to force usage (may lead to runtime failures)

**Review by ruskcoder — COMMENTED** [review]
I fixed all the files, and added a few bug fixes. My Android Studio also automatically upgraded AGP to 8.5.1.

**Comment by ruskcoder** [comment]
Is this going to be implemented, or are there things I need to fix?

File: app/build.gradle.kts [hunk 17]
```diff
@@ -28,6 +28,7 @@ android {
             isMinifyEnabled = true
             isShrinkResources = true
             proguardFiles(getDefaultProguardFile("proguard-android-optimize.txt"), "proguard-rules.pro")
+            signingConfig = signingConfigs.getByName("debug")
```
Comment (mikooomich) [inline app/build.gradle.kts:31]: remove this

File: app/proguard-rules.pro [hunk 18]
```diff
@@ -58,4 +69,4 @@
 -dontwarn org.openjsse.javax.net.ssl.SSLParameters
 -dontwarn org.openjsse.javax.net.ssl.SSLSocket
 -dontwarn org.openjsse.net.ssl.OpenJSSE
--dontwarn org.slf4j.impl.StaticLoggerBinder
+-dontwarn org.slf4j.impl.StaticLoggerBinder
```
Comment (mikooomich) [inline app/proguard-rules.pro:72]: revert this change

File: app/src/main/java/com/dd3boh/outertune/playback/MusicService.kt
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/MusicService.kt:1]: these 2 changes are unnecessary

File: app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt:1]: it's probably alot nicer for all this to go inside MusicService instead. You'd have access to context and datastore easily; avoids passing context from mainactivity

File: app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt [hunk 19]
```diff
@@ -109,17 +119,39 @@ class PlayerConnection(
     override fun onPlaybackStateChanged(state: Int) {
         playbackState.value = state
         error.value = player.playerError
+
+        if (state == 2) {
+            closeDiscordRPC(ctx = ctx)
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt:124]: Use state enums plz.

ex. STATE_READY

The current impl is a bit "flaky, idk where exactly the issue is
- dragging the seek bar recreates rpc even if its same song
- manually skipping/click to play does not load new song
- onMediaItemTransition should be all you really need iirc (for the non-pause impl)

File: app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt [hunk 20]
```diff
@@ -109,17 +119,39 @@ class PlayerConnection(
...
+            }
+        }
     }
 
     override fun onMediaItemTransition(mediaItem: MediaItem?, reason: Int) {
         mediaMetadata.value = mediaItem?.metadata
         currentMediaItemIndex.value = player.currentMediaItemIndex
         currentWindowIndex.value = player.getCurrentQueueIndex()
         updateCanSkipPreviousAndNext()
+        if (reason == 1) {
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/PlayerConnection.kt:147]: use enums plz

ex. MEDIA_ITEM_TRANSITION_REASON_AUTO


Why only reason 1 here? It only recreates RPC when auto skip, and not for when user seeks?  (add code comment with why if this is intended)

File: app/src/main/java/com/dd3boh/outertune/ui/screens/settings/DiscordLoginScreen.kt
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/ui/screens/settings/DiscordLoginScreen.kt:1]: use single indents only

File: build.gradle.kts
Comment (mikooomich) [inline build.gradle.kts:1]: remove changes

File: gradle.properties
Comment (mikooomich) [inline gradle.properties:1]: remove changes

File: gradle/libs.versions.toml [hunk 21]
```diff
@@ -70,6 +72,8 @@ junit = { group = "junit", name = "junit", version = "4.13.2" }
 
 timber = { group = "com.jakewharton.timber", name = "timber", version = "5.0.1" }
 
+gson = { group = "com.google.code.gson", name = "gson", version = "2.11.0" }
+
```
Comment (mikooomich) [inline gradle/libs.versions.toml:76]: what is gson for? kizzyrpc dependency?

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+
+                } catch (e: Exception) {
+                    reportException(e)
+                } finally {
+                    client.close()
+                }
+            }
+        }
+
+        else{
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt:131]: else{ -> else {

File: app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+                        } finally {
+                            clientDiscordCDN.close()
+                        }
+                    }
+
+                    CoroutineScope(Dispatchers.IO).launch {
+                        try {
+                            val response: HttpResponse = clientDiscordCDN.get("https://kizzyapi-1-z9614716.deta.app/image?url=" + artwork)
+                            artworkCDN = JSONObject(response.bodyAsText()).getString("id")
+                            Thread.sleep(500)
```
Comment (mikooomich) [inline app/src/main/java/com/dd3boh/outertune/playback/DiscordRPCInfoFetch.kt:97]: waiting for response to return? Pretty dangerous given the uncertainty that is networking

**Review by mikooomich — CHANGES_REQUESTED** [review]
Sorry for the delay, had some things to attend to. 

- album arts does not work with local songs (Honestly a placeholder image or even no image is fine)
- "Listening to <title>" should be "Listening to OuterTune"  (there is a string.xml entry called app_name you can use)
- Move Discord settings entry should be under content settings, account heading. probably a title like Discord RPC or Discord Rich Presence would be a better name
- Imo "unstable" features should be off by default

If you could clean up the commits a bit that would be nice. (for readability sakes) It's probably best to make any changes, and then squash everything, then Split into multiple smaller commits if you wish.

**Comment by ruskcoder** [comment]
Actually, I also started working on the original Innertune project and added a full fledged and completed discord (cleaned up and everything). When I finish there, I will just copy the commits over so it is the same.

**Comment by mikooomich** [comment]
Ah cool! To save you time, base off of release/0.6.0, not the ancient dev branch

**Comment by ruskcoder** [comment]
The Discord RPC thing for the official InnerTune was completed today, by next week (my school started and I'm quite busy right now), I will try to merge it into OuterTune.
